### PR TITLE
[nG] Gather-7 update for negative batch_dims

### PR DESCRIPTION
**Pull request by pavel-esir**
### Details:
 - added fixed `batch_dims` normalization, this led to some changes in shape validation. Now negative `batch_dims` are processed same as in TF https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/gather_op.cc#L100
- corrected a typo in specification and added clarification that indices can be also scalar
- reverted back Gather-1 public methods for source compatibility with older ngraph library versions:
   - removed  from v1 get_batch_dims, is_axis_set
   - returned back `AXIS_NOT_SET_VALUE`
   - marked `get_axis` as virtual and overloaded in v0:Gather to mimic old v0 behavior

### Tickets:
 - #54329

### CODE REVIEW

File: ngraph/core/src/op/util/gather_base.cpp [hunk 1]
```diff
@@ -52,26 +52,26 @@ void op::util::GatherBase::validate_and_infer_types()
             axis_pshape);
     }
 
-    int64_t batch_dims = get_batch_dims(); // will not be converted to positive if axis is not set
+    int64_t batch_dims = get_batch_dims(); // will not be normalized if indices rank is dynamic
     if (is_axis_set())
     {
         int64_t axis = get_axis();
+
+        // batch_dims, axis both can e positive by default or if data_rank & indices_rank are
```
Comment (achetver) [inline ngraph/core/src/op/util/gather_base.cpp:60]: Please, fix typo - `...both can be positive by...` maybe?.

Comment (pavel-esir) [reply ngraph/core/src/op/util/gather_base.cpp:60]: will fix this together with other comments which I'm sure other will also leave

Comment (pavel-esir) [reply ngraph/core/src/op/util/gather_base.cpp:60]: fixed

File: inference-engine/src/transformations/src/transformations/common_optimizations/algebraic_simplification.cpp [hunk 2]
```diff
@@ -37,12 +37,11 @@ static bool simplify_gather(std::shared_ptr<Node> node) {
             return false;
         }
 
-
-        if (!gather->is_axis_set()) {
+        auto axis = gather->get_axis();
+        if (axis == opset3::Gather::AXIS_NOT_SET_VALUE) {
```
Comment (ilyachur) [inline inference-engine/src/transformations/src/transformations/common_optimizations/algebraic_simplification.cpp:41]: Am I right that you have changed the public API of Gather from opset3 and removed public method `is_axis_set()`?

Comment (pavel-esir) [reply inference-engine/src/transformations/src/transformations/common_optimizations/algebraic_simplification.cpp:41]: No, previously when I was enabling Gather-7 I added `is_axis_set()` https://github.com/openvinotoolkit/openvino/pull/4885/files#diff-1899273196bcbed10cce05f78cfe5281ab7b190af61e39452f4ab6eb7c774c28R82 
and now reverted back to `AXIS_NOT_SET_VALUE` to match old public API.

Here are status far before my changes for Gather-7, `AXIS_NOT_SET_VALUE` was present there
https://github.com/openvinotoolkit/openvino/blob/35bff0d8a1e4c7fb6ab55ff8948a9bc1019b218c/inference-engine/src/transformations/src/transformations/common_optimizations/algebraic_simplification.cpp#L52-L57

https://github.com/openvinotoolkit/openvino/blob/35bff0d8a1e4c7fb6ab55ff8948a9bc1019b218c/ngraph/core/src/op/gather.cpp#L34

https://github.com/openvinotoolkit/openvino/blob/35bff0d8a1e4c7fb6ab55ff8948a9bc1019b218c/ngraph/core/include/ngraph/op/gather.hpp#L31

I personally think that `is_axis_set()` is better but if we cannot change public API then I do not insist on solution with `is_axis_set()`

**Review by mvafin — APPROVED** [review]

File: ngraph/core/src/op/util/gather_base.cpp [hunk 3]
```diff
@@ -52,26 +52,36 @@ void op::util::GatherBase::validate_and_infer_types()
...
-            NODE_VALIDATION_CHECK(this,
-                                  axis >= 0 && axis < data_rank.get_length(),
-                                  "The axis must be >= 0 and < data_rank. But instead got axis = ",
-                                  axis,
-                                  " data_rank = ",
-                                  data_rank.get_length());
-        }
+        NODE_VALIDATION_CHECK(this,
+                              axis < data_rank.get_length() || data_rank.is_dynamic(),
+                              "The axis must be >= 0 and < data_rank. But instead got axis = ",
```
Comment (popovaan) [inline ngraph/core/src/op/util/gather_base.cpp:81]: Please, update the message to correspond the check.

**Review by popovaan — APPROVED** [review]

File: ngraph/core/src/op/util/gather_base.cpp (hunk 3 above)
Comment (pavel-esir) [reply ngraph/core/src/op/util/gather_base.cpp:82]: thanks! fixed

**Review by tsavina — APPROVED** [review]

**Review by achetver — APPROVED** [review]

**Review by ilyachur — COMMENTED** [review]
Am I right that you have reverted all backward incompatible changes from gather v3?

**Comment by pavel-esir** [comment]
> Am I right that you have reverted all backward incompatible changes from gather v3?

@ilyachur yes, but from gather v1 (but opset3::Gather includes v1::Gather)

File: ngraph/core/src/op/util/gather_base.cpp [hunk 4]
```diff
@@ -52,26 +52,37 @@ void op::util::GatherBase::validate_and_infer_types()
...
     {
         int64_t axis = get_axis();
+
+        // batch_dims, axis both can be positive by default or after normalization if data_rank &
+        // indices_rank are static. If at least one of them is negative we cannot check their
+        // consistency.
         NODE_VALIDATION_CHECK(this,
-                              batch_dims <= axis,
+                              batch_dims <= axis || batch_dims < 0 || axis < 0,
                               "The batch_dims <= axis. But instead got: batch_dims = ",
```
Comment (ilyachur) [inline ngraph/core/src/op/util/gather_base.cpp:74]: nitpick: we have incorrect error message 😄 because negative `batch_dims` can be `<=` than axis 😄

**Review by ilyachur — APPROVED** [review]
Please fix error messages in the next PR

**Pull request merged**